### ISSUE TITLE
Performance: the newrelic-telemetry-sdk-go uses too much memory and CPU when heavily loaded

### ISSUE THREAD

**Issue by justinfoote** [newrelic/newrelic-telemetry-sdk-go#37]
See also: https://github.com/aws-observability/aws-otel-collector/issues/180

When the New Relic Telemetry SDK receives a moderate number of traces (around 100 per second), the CPU and memory use are not exceptionally high, but as the number of traces per second increase, the memory and CPU overhead become extremely high

From [AWS's testing](https://github.com/aws-observability/aws-otel-collector/blob/df9e1e418ffc78339740eaff56132f0da02d2100/docs/performance_model.md):
 | Test Case | Data Rate(tps) |  Data Type | Instance Type | Avg CPU Usage (Percent Per Core) | Avg Memory Usage (Megabytes) | 
 |:---------:|:---------:|:----------:|:------------:|:-----------------------:|:----------------------------:| 
 |newrelic_exporter_trace_mock | 100 | otlp | m5.2xlarge | 23.10 | 65.86 || 
 |newrelic_exporter_trace_mock | 1000 | otlp | m5.2xlarge | 386.52 | 4294.13 || 
 |newrelic_exporter_trace_mock | 5000 | otlp | m5.2xlarge | 553.62 | 22009.65 || 

From my own testing:
```json
[
  {
    "avgCpu": 23.31207055599699,
    "avgMem": 64.9428992,
    "collectionPeriod": 10,
    "dataRate": 100,
    "dataType": "otlp",
    "instanceType": "m5.2xlarge",
    "testcase": "newrelic_exporter_trace_mock",
    "testingAmi": "soaking_linux"
  },
  {
    "avgCpu": 599.9423109889208,
    "avgMem": 19867.21252693333,
    "collectionPeriod": 10,
    "dataRate": 5000,
    "dataType": "otlp",
    "instanceType": "m5.2xlarge",
    "testcase": "newrelic_exporter_trace_mock",
    "testingAmi": "soaking_linux"
  }
]
```

**Comment by justinfoote** [comment]
I believe that most of this performance overhead is to deal with a restriction coming from New Relic ingest. If the Telemetry SDK sends a payload larger that 1MB (compressed), the New Relic backend will respond with a 413. According to [the spec](https://github.com/newrelic/newrelic-telemetry-sdk-specs/blob/master/communication.md#splitting-data), the SDK should deal with that 413 by splitting the payload in half and trying again. 

The spec allows an SDK to _preemptively_ split the payloads though -- and this is exactly what the go telemetry SDK does.
In the [`newRequestsInternal` function](https://github.com/newrelic/newrelic-telemetry-sdk-go/blob/master/telemetry/request.go#L45) in request.go, a payload is compressed, then its size is compared to a hardcoded max size (1MB). If the compressed payload is larger than 1MB, the payload is split in half, then the `newRequestsInternal` function is called, recursively, on each of the resulting payloads. 
While this recursion is happening, the original payload is kept in memory as:
* an array of structs
* an uncompressed json string
* a compressed bytearray
* a compressed string

If a payload is extremely large (which may happen when the SDK is extremely overloaded), the payloads will require more than one split. In that case, each request will be compressed several times, and every interim payload will be kept in memory until all of the splitting is done.

**Comment by justinfoote** [comment]
There is some low-hanging fruit here.
* We could de-scope the request before recursing -- this would free up quite a lot of memory.
* Rather than splitting in half, we could compute how many buckets a payload should be split into -- this would (in most cases) allow us to split and send payloads with at most one level of recursion.

**Comment by a-feld** [comment]
A lot of this boiled down to allocating a new gzip compressor for each payload. This has been fixed by #44 
Based on performance testing under load, it's likely performance is now acceptable so we can close this issue for now. If there are additional issues that come up, we can reopen this issue.
